src/pcre2_util.h

```diff
@@ -44,26 +44,52 @@ POSSIBILITY OF SUCH DAMAGE.
 /* Assertion macros */
 
 #ifdef HAVE_BUILTIN_UNREACHABLE
-#define PCRE2_UNREACHABLE() __builtin_unreachable()
-#elif HAVE_BUILTIN_ASSUME
-#define PCRE2_UNREACHABLE() __assume(0)
+#define PCRE2_DEADTRAP() __builtin_unreachable()
+#elif defined(HAVE_BUILTIN_ASSUME)
+#define PCRE2_DEADTRAP() __assume(0)
 #else
-#define PCRE2_UNREACHABLE() do {} while(0)
+#warning "missing implementation of PCRE2_UNREACHABLE()"
 #endif
 
 #ifdef PCRE2_DEBUG
+
 #if defined(HAVE_ASSERT_H) && !defined(NDEBUG)
 #include <assert.h>
 #define PCRE2_ASSERT(x) assert(x)
 #elif defined(HAVE_STDLIB_H) && defined(HAVE_STDIO_H)
-#define PCRE2_ASSERT(x) do { if (!(x)) { fprintf(stderr, "Assertion failed at " __FILE__ ":%d\n", __LINE__); abort(); }} while(0)
+#define PCRE2_ASSERT(x) do                                            \
+{                                                                     \
+  if (!(x))                                                           \
+  {                                                                   \
+  fprintf(stderr, "Assertion failed at " __FILE__ ":%d\n", __LINE__); \
+  abort();                                                            \
+  }                                                                   \
+} while(0)
+#else
+#warning "missing implementation of PCRE2_ASSERT()"
 #endif
+
+#define PCRE2_UNREACHABLE(d) do                                     \
+{                                                                   \
+fprintf(stderr, "Execution reached unexpected point " __FILE__      \
+                ":%d\n", __LINE__);                                 \
+abort();                                                            \
+} while(0)
+
+#endif /* PCRE2_DEBUG */
+
+#ifndef PCRE2_UNREACHABLE
+#ifdef PCRE2_DEADTRAP
+#define PCRE2_UNREACHABLE(d) if (d == 0) PCRE2_DEADTRAP()
```

Comment:
Ah, I see. So the parameter to PCRE2_UNREACHABLE indicates whether we have confidence that the code is "really unreachable" or not.

Interesting.

Reply:
Guess that is a way to put it, but what it really indicates to me is, "this unreachable should NEVER abort in non debug builds and should NEVER be considered as a hint for code elimination either, it is there ONLY to be used in debug builds to catch bugs, as all assertions should be.

Reply:
Instead of a parameter, I would introduce two unreachable variants. PCRE2_UNREACHABLE / PCRE2_SOFT_UNREACHABLE (or any better name).

Reply:
The names are good. If I understand correctly any unreachable is a bug, just we have a fallback in the latter case.